#!/usr/bin/env python3
# client/calculator_client.py

import grpc
import time
import sys
import os
from datetime import datetime

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from generated import calculator_pb2
from generated import calculator_pb2_grpc


class CalculatorClient:
    """è®¡ç®—å™¨gRPCå®¢æˆ·ç«¯"""

    def __init__(self, host='localhost', port=50051):
        """åˆå§‹åŒ–å®¢æˆ·ç«¯è¿æ¥"""
        self.channel = grpc.insecure_channel(f'{host}:{port}')
        self.stub = calculator_pb2_grpc.CalculatorStub(self.channel)
        print(f"âœ… å·²è¿æ¥åˆ°æœåŠ¡å™¨ {host}:{port}")

    def add(self, num1, num2):
        """ä¸€å…ƒRPCï¼šç®€å•åŠ æ³•"""
        print(f"\n{'=' * 50}")
        print("ğŸ“¤ è°ƒç”¨ç®€å•åŠ æ³• (ä¸€å…ƒRPC)")
        print(f"{'=' * 50}")

        request = calculator_pb2.AddRequest(
            num1=num1,
            num2=num2,
            request_id=f"req_{int(time.time())}"
        )

        try:
            start_time = time.time()
            response = self.stub.Add(request)
            elapsed = (time.time() - start_time) * 1000

            print(f"è¯·æ±‚: {num1} + {num2}")
            print(f"å“åº”: {response.message}")
            print(f"ç»“æœ: {response.result}")
            print(f"æ—¶é—´: {datetime.fromtimestamp(response.timestamp / 1000).strftime('%Y-%m-%d %H:%M:%S')}")
            print(f"è€—æ—¶: {elapsed:.2f}ms")

            return response.result

        except grpc.RpcError as e:
            print(f"âŒ RPCè°ƒç”¨å¤±è´¥: {e.code()} - {e.details()}")
            return None

    def add_stream(self, num1, num2):
        """æœåŠ¡ç«¯æµå¼RPCï¼šè·å–è®¡ç®—æ­¥éª¤"""
        print(f"\n{'=' * 50}")
        print("ğŸ“¤ è°ƒç”¨æµå¼åŠ æ³• (æœåŠ¡ç«¯æµRPC)")
        print(f"{'=' * 50}")

        request = calculator_pb2.AddRequest(
            num1=num1,
            num2=num2,
            request_id=f"stream_req_{int(time.time())}"
        )

        try:
            start_time = time.time()

            print("å¼€å§‹æ¥æ”¶æµå¼å“åº”...")
            for i, step in enumerate(self.stub.AddStream(request), 1):
                print(f"\næ­¥éª¤ {step.step}:")
                print(f"  {step.message}")
                print(f"  å½“å‰ç»“æœ: {step.current_result}")
                print(f"  æ—¶é—´: {datetime.fromtimestamp(step.timestamp / 1000).strftime('%H:%M:%S.%f')[:-3]}")

            elapsed = (time.time() - start_time) * 1000
            print(f"\nâœ“ æµå¼è¯·æ±‚å®Œæˆï¼Œæ€»è€—æ—¶: {elapsed:.2f}ms")

        except grpc.RpcError as e:
            print(f"âŒ RPCè°ƒç”¨å¤±è´¥: {e.code()} - {e.details()}")

    def add_client_stream(self, numbers):
        """å®¢æˆ·ç«¯æµå¼RPCï¼šå‘é€å¤šä¸ªæ•°å­—è¿›è¡Œç´¯åŠ """
        print(f"\n{'=' * 50}")
        print("ğŸ“¤ è°ƒç”¨å®¢æˆ·ç«¯æµå¼ç´¯åŠ  (å®¢æˆ·ç«¯æµRPC)")
        print(f"{'=' * 50}")

        def number_generator():
            """ç”Ÿæˆæ•°å­—çš„ç”Ÿæˆå™¨"""
            for i, num in enumerate(numbers, 1):
                print(f"å‘é€ç¬¬{i}ä¸ªæ•°å­—: {num}")
                yield calculator_pb2.Number(value=num)
                time.sleep(0.3)  # æ¨¡æ‹Ÿç”¨æˆ·è¾“å…¥é—´éš”

        try:
            start_time = time.time()
            response = self.stub.AddClientStream(number_generator())
            elapsed = (time.time() - start_time) * 1000

            print(f"\nâœ“ æœåŠ¡å™¨å“åº”:")
            print(f"  {response.message}")
            print(f"  æ€»å’Œ: {response.result}")
            print(f"  è¯·æ±‚ID: {response.request_id}")
            print(f"  æ€»è€—æ—¶: {elapsed:.2f}ms")

            return response.result

        except grpc.RpcError as e:
            print(f"âŒ RPCè°ƒç”¨å¤±è´¥: {e.code()} - {e.details()}")
            return None

    def add_bidirectional(self, requests):
        """åŒå‘æµå¼RPCï¼šå®æ—¶äº¤äº’è®¡ç®—"""
        print(f"\n{'=' * 50}")
        print("ğŸ“¤ è°ƒç”¨åŒå‘æµå¼è®¡ç®— (åŒå‘æµRPC)")
        print(f"{'=' * 50}")

        def request_generator():
            """ç”Ÿæˆè¯·æ±‚çš„ç”Ÿæˆå™¨"""
            for i, (num1, num2) in enumerate(requests, 1):
                request = calculator_pb2.AddRequest(
                    num1=num1,
                    num2=num2,
                    request_id=f"bidir_{i}_{int(time.time())}"
                )
                print(f"å‘é€è¯·æ±‚ {i}: {num1} + {num2}")
                yield request
                time.sleep(0.5)  # æ¨¡æ‹Ÿè¯·æ±‚é—´éš”

        try:
            start_time = time.time()
            print("å¼€å§‹åŒå‘æµé€šä¿¡...\n")

            for i, response in enumerate(self.stub.AddBidirectional(request_generator()), 1):
                print(f"æ”¶åˆ°ç¬¬{i}ä¸ªå“åº”:")
                print(f"  {response.message}")
                print(f"  æ­¥éª¤: {response.step}, ç»“æœ: {response.current_result}")
                print(f"  æ—¶é—´: {datetime.fromtimestamp(response.timestamp / 1000).strftime('%H:%M:%S.%f')[:-3]}")
                print()

            elapsed = (time.time() - start_time) * 1000
            print(f"âœ“ åŒå‘æµé€šä¿¡å®Œæˆï¼Œæ€»è€—æ—¶: {elapsed:.2f}ms")

        except grpc.RpcError as e:
            print(f"âŒ RPCè°ƒç”¨å¤±è´¥: {e.code()} - {e.details()}")

    def close(self):
        """å…³é—­è¿æ¥"""
        self.channel.close()
        print("è¿æ¥å·²å…³é—­")


def main():
    """ä¸»å‡½æ•°ï¼šæ¼”ç¤ºæ‰€æœ‰RPCè°ƒç”¨"""
    # åˆ›å»ºå®¢æˆ·ç«¯
    client = CalculatorClient()

    try:
        # 1. ç®€å•åŠ æ³•
        print("\n" + "ğŸ¯ æ¼”ç¤º1: ç®€å•åŠ æ³•".center(50, "="))
        result1 = client.add(3.5, 7.2)
        result2 = client.add(100, 200)

        # 2. æœåŠ¡ç«¯æµå¼åŠ æ³•
        print("\n" + "ğŸ¯ æ¼”ç¤º2: æœåŠ¡ç«¯æµå¼åŠ æ³•".center(50, "="))
        client.add_stream(15, 25)

        # 3. å®¢æˆ·ç«¯æµå¼ç´¯åŠ 
        print("\n" + "ğŸ¯ æ¼”ç¤º3: å®¢æˆ·ç«¯æµå¼ç´¯åŠ ".center(50, "="))
        numbers_to_send = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
        client.add_client_stream(numbers_to_send)

        # 4. åŒå‘æµå¼è®¡ç®—
        print("\n" + "ğŸ¯ æ¼”ç¤º4: åŒå‘æµå¼è®¡ç®—".center(50, "="))
        requests = [(1, 2), (3, 4), (5, 6), (10, 20)]
        client.add_bidirectional(requests)

        # 5. é”™è¯¯å¤„ç†æ¼”ç¤º
        print("\n" + "ğŸ¯ æ¼”ç¤º5: é”™è¯¯å¤„ç†".center(50, "="))

        # å°è¯•ä¸ä¸å­˜åœ¨çš„æœåŠ¡å™¨é€šä¿¡
        print("\nå°è¯•è¿æ¥ä¸å­˜åœ¨çš„æœåŠ¡å™¨...")
        try:
            broken_client = CalculatorClient('localhost', 9999)
            broken_client.add(1, 2)
        except Exception as e:
            print(f"é¢„æœŸä¸­çš„è¿æ¥å¤±è´¥: {type(e).__name__}")

    except KeyboardInterrupt:
        print("\n\nğŸ‘‹ ç”¨æˆ·ä¸­æ–­")
    finally:
        # å…³é—­è¿æ¥
        client.close()


if __name__ == '__main__':
    main()